// @base44/vite-plugin/src/injections/utils.ts
/** Check if an element has instrumentation attributes */
export function isInstrumentedElement(element: Element): boolean {
  const htmlEl = element as HTMLElement;
  return !!(
    htmlEl.dataset?.sourceLocation || htmlEl.dataset?.visualSelectorId
  );
}

/** Get the selector ID from an element's data attributes (prefers source-location) */
export function getElementSelectorId(element: Element): string | null {
  const htmlEl = element as HTMLElement;
  return (
    htmlEl.dataset?.sourceLocation ||
    htmlEl.dataset?.visualSelectorId ||
    null
  );
}

export const ALLOWED_ATTRIBUTES: string[] = ["src"];

/** Find elements by ID - first try data-source-location, fallback to data-visual-selector-id */
export function findElementsById(id: string | null): Element[] {
  if (!id) return [];
  const sourceElements = Array.from(
    document.querySelectorAll(`[data-source-location="${id}"]`)
  );
  if (sourceElements.length > 0) {
    return sourceElements;
  }
  return Array.from(
    document.querySelectorAll(`[data-visual-selector-id="${id}"]`)
  );
}

/**
 * Update element classes by visual selector ID.
 * Uses setAttribute instead of className to support both HTML and SVG elements.
 */
export function updateElementClasses(elements: Element[], classes: string): void {
  elements.forEach((element) => {
    element.setAttribute("class", classes);
  });
}

/** Set a single attribute on all provided elements. */
export function updateElementAttribute(elements: Element[], attribute: string, value: string): void {
  if (!ALLOWED_ATTRIBUTES.includes(attribute)) {
    return;
  }

  elements.forEach((element) => {
    element.setAttribute(attribute, value);
  });
}

/** Collect attribute values from an element for a given allowlist. */
export function collectAllowedAttributes(element: Element, allowedAttributes: string[]): Record<string, string> {
  const attributes: Record<string, string> = {};
  for (const attr of allowedAttributes) {
    const val = element.getAttribute(attr);
    if (val !== null) {
      attributes[attr] = val;
    }
  }
  return attributes;
}
