// @base44/vite-plugin/src/injections/layer-dropdown/consts.ts
/** Style constants for the layer dropdown UI */

export const DROPDOWN_CONTAINER_STYLES: Record<string, string> = {
  position: "absolute",
  backgroundColor: "#ffffff",
  border: "1px solid #e2e8f0",
  borderRadius: "6px",
  boxShadow: "0 4px 12px rgba(0, 0, 0, 0.15)",
  fontSize: "12px",
  minWidth: "120px",
  maxHeight: "200px",
  overflowY: "auto",
  zIndex: "10001",
  padding: "4px 0",
  pointerEvents: "auto",
};

export const DROPDOWN_ITEM_BASE_STYLES: Record<string, string> = {
  padding: "4px 12px",
  cursor: "pointer",
  color: "#334155",
  backgroundColor: "transparent",
  whiteSpace: "nowrap",
  lineHeight: "1.5",
  fontWeight: "400",
};

export const DROPDOWN_ITEM_ACTIVE_COLOR = "#526cff";
export const DROPDOWN_ITEM_ACTIVE_BG = "#DBEAFE";
export const DROPDOWN_ITEM_ACTIVE_FONT_WEIGHT = "600";

export const DROPDOWN_ITEM_HOVER_BG = "#f1f5f9";

export const DEPTH_INDENT_PX = 10;

/** Chevron shown when dropdown is collapsed (click to expand) */
export const CHEVRON_COLLAPSED = " \u25BE";
/** Chevron shown when dropdown is expanded (click to collapse) */
export const CHEVRON_EXPANDED = " \u25B4";

export const BASE_PADDING_PX = 12;

export const LAYER_DROPDOWN_ATTR = "data-layer-dropdown";

/** Max instrumented ancestors to show above the selected element */
export const MAX_PARENT_DEPTH = 2;

/** Max instrumented depth levels to show below the selected element */
export const MAX_CHILD_DEPTH = 2;
